internal/transport/http2_client.go

```diff
@@ -1309,6 +1312,7 @@ func (t *http2Client) reader() {
 	// Check the validity of server preface.
 	frame, err := t.framer.fr.ReadFrame()
 	if err != nil {
+		t.updateConnectionError(err)
```

Comment:
One thing about my design in #4163 that I preferred is that adding an `error` to `Close` means we're forced to inspect every place it is called (now and in the future) so we can improve our error reporting in *all* instances where the transport is closed.  With this alternate implementation, we only *optionally* improve it.  I'm fine with the `LastConnectionError()` method (though it is stateful, and, as such, I do prefer the alternative of making `OnClose` accept an error result).  But we should really make `http2Client.Close` accept an `error`.  If there is no useful error at the site where `Close` is called, just pass `ErrConnClosing`.

If we do that, we also know that whenever `OnClose` is called, there *must* be a non-nil `error`, and we can simplify the `clientconn.go` code to always return `nil, nil, t.LastConnectionError()` (without the nil check).

Reply:
Ok, I will go with your idea. Should I keep LastConnectionError or have onClose callback to receive the connection error as parameter?

Reply:
`LastConnectionError` is stateful, meaning it requires synchronization and opens the possibility for a new class of bugs.  I think if `OnClose` takes an error to pass the information, it makes for a better flow.  (This is "share memory by communicating instead of communicating by sharing memory," it just doesn't require a channel.)